plugins/palo_alto_cortex_xdr/icon_palo_alto_cortex_xdr/tasks/monitor_alerts/task.py

```diff
@@ -0,0 +1,381 @@
+
+                self.logger.info(
+                    f"Found total alerts={total_count}, limit={ALERT_LIMIT}, is_paginating={is_paginating}"
+                )
+                self.logger.info(
+                    f"Paginating alerts: Saving state with existing filters: "
+                    f"from_time={search_from}"
+                    f"to_time={search_to}"
+                )
+
+            else:
+                has_more_pages = False
+                state = self._drop_pagination_state(state)
+
+            return results, state, has_more_pages
+
+        except PluginException as error:
+            self.logger.error(
+                f"A PluginException has occurred. Status code 500 returned. Error: {error.cause}. "
+                f"Existing state: {state}"
+            )
+            return [], state, False, 500, PluginException(cause=error.cause, data=error)
+
+        except Exception as error:
+            print(f"{error = }")
+            return [], state, False, 500, PluginException(preset=PluginException.Preset.UNKNOWN, data=error)
+
+    ###########################
+    # Deduping
+    ###########################
+    def _dedupe_and_get_highest_time(self, alerts: list, start_time: str, state: dict) -> Tuple[list, list, str]:
+        """
+        Function to dedupe alerts using existing hashes in the state, and return the highest
+        timestamp.
+
+        :param alerts: list of alerts
+        :param state: state of the plugin
+        :return: list of deduped alerts, list of new hashes, and the highest timestamp
+        """
+
+        old_hashes, deduped_alerts, new_hashes, highest_timestamp = state.get(LAST_ALERT_HASH, []), [], [], ""
+        for index, alert in enumerate(alerts):
+
+            # todo - make method to convert
+            alert_time = alert.get(TIMESTAMP_KEY) / 1000
+            alert_time = datetime.fromtimestamp(alert_time, tz=timezone.utc).strftime(TIME_FORMAT)
+
+            if alert_time == start_time:
+                alert_hash = hash_sha1(alert)
+                if alert_hash not in old_hashes:
+                    deduped_alerts.append(alert)
+            elif alert_time > start_time:
+                deduped_alerts += alerts[index:]
+                print(f"{deduped_alerts= }")
+                break
+
+        num_alerts = len(deduped_alerts)
+        self.logger.info(f"Received {len(alerts)} alerts, and after dedupe there is {num_alerts} results.")
+
+        if deduped_alerts:
+            # alert results are already sorted by CS API, so get the latest timestamp
+            highest_timestamp = deduped_alerts[-1].get(TIMESTAMP_KEY)
+
+            for alert in deduped_alerts:
+                if alert.get(TIMESTAMP_KEY) == highest_timestamp:
+                    new_hashes.append(hash_sha1(alert))
+                    self.logger.info(f"Hashed latest event with timestamp {highest_timestamp}.")
+
+            self.logger.debug(f"Highest timestamp is {highest_timestamp}")
```

Comment:
We'll need to set this highest timestamp in state to use in future queries if we are no longer paginating but are catching up.